libraries/ml/effekt.effekt

```diff
@@ -196,24 +196,23 @@ extern io def panic[R](msg: String): R =
   "raise Fail msg"
 
 effect Exception[E] {
-  def raise(exception: E, msg: String): Nothing
+  def raise[A](exception: E, msg: String): A
 }
```

Comment:
This is a notable change since all of the other backends used `def raise[A](exception: E, msg: String): A`, but the ML backend used `def raise(exception: E, msg: String): Nothing`.
What's the better choice?

Reply:
That's a very good question. The only reason we do this in MLton is since it cannot monomorphize rank-2. This will not be the last rank-2 operation, so we need a general strategy for dealing with this.

Reply:
Although thinking about the possible backends (and compilation strategies) is important, I feel like the question here is a semantic one: what is the _ideal_ version of `raise`? Returning A for all A or returning Nothing? Which variant is nicer/easier to work with?

Reply:
Of course in terms of parametricity, both are the same. For this particular case I might actually prefer `Nothing` since it is more explicit and less indirect. We need to make sure that the subtyping has the right ergonomics in the concrete implementation, though.